[BUG] 'sparkDriver' Error by running MLflow server for spark saved model

Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [x] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MAC OS 
- **MLflow installed from (source or binary)**: pip install
- **MLflow version (run ``mlflow --version``)**: 1.25.1
- **Python version**: 3.8.12
- **npm version, if running the dev UI**: 8.5.0
- **Exact command to reproduce**: ```mlflow models serve -m spark_model --no-conda --host 0.0.0.0 --port 15```

### Describe the problem

when i try to run this command in terminal mlflow models serve -m spark_model --no-conda --host 0.0.0.0 --port 15
and getting this as error
Can't assign requested address: Service 'sparkDriver' failed after 16 retries (on a random free port)

spark_model was saved using mlflow.spark.save_model()

Would like to know the solution for this, Thanks !

Below is the whole error:

(spark) i29140@LVVCHD7QMD6R spark % mlflow models serve -m spark_model --no-conda --host 0.0.0.0 --port 15
/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py:2322: FutureWarning: `--no-conda` is deprecated and will be removed in a future MLflow release. Use `--env-manager=local` instead.
  value = self.callback(ctx, self, value)
2022/04/20 09:43:11 INFO mlflow.models.cli: Selected backend for flavor 'python_function'
2022/04/20 09:43:11 INFO mlflow.pyfunc.backend: === Running command 'gunicorn --timeout=60 -b 0.0.0.0:15 -w 1 ${GUNICORN_CMD_ARGS} -- mlflow.pyfunc.scoring_server.wsgi:app'
[2022-04-20 09:43:11 -0700] [1359] [INFO] Starting gunicorn 20.1.0
[2022-04-20 09:43:11 -0700] [1359] [INFO] Listening at: http://0.0.0.0:15 (1359)
[2022-04-20 09:43:11 -0700] [1359] [INFO] Using worker: sync
[2022-04-20 09:43:11 -0700] [1361] [INFO] Booting worker with pid: 1361
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/04/20 09:43:14 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
22/04/20 09:43:14 ERROR SparkContext: Error initializing SparkContext.
java.net.BindException: Can't assign requested address: Service 'sparkDriver' failed after 16 retries (on a random free port)! Consider explicitly setting the appropriate binding address for the service 'sparkDriver' (for example spark.driver.bindAddress for SparkDriver) to the correct binding address.
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:438)
	at sun.nio.ch.Net.bind(Net.java:430)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:225)
	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:562)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:260)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)
[2022-04-20 09:43:14 -0700] [1361] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/gunicorn/util.py", line 359, in import_app
    mod = importlib.import_module(module)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/mlflow/pyfunc/scoring_server/wsgi.py", line 6, in <module>
    app = scoring_server.init(load_model(os.environ[scoring_server._SERVER_MODEL_PATH]))
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 733, in load_model
    model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/mlflow/spark.py", line 729, in _load_pyfunc
    pyspark.sql.SparkSession.builder.config("spark.python.worker.reuse", True)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/pyspark/sql/session.py", line 228, in getOrCreate
    sc = SparkContext.getOrCreate(sparkConf)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/pyspark/context.py", line 392, in getOrCreate
    SparkContext(conf=conf or SparkConf())
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/pyspark/context.py", line 146, in __init__
    self._do_init(master, appName, sparkHome, pyFiles, environment, batchSize, serializer,
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/pyspark/context.py", line 209, in _do_init
    self._jsc = jsc or self._initialize_context(self._conf._jconf)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/pyspark/context.py", line 329, in _initialize_context
    return self._jvm.JavaSparkContext(jconf)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/py4j/java_gateway.py", line 1585, in __call__
    return_value = get_return_value(
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/py4j/protocol.py", line 326, in get_return_value
    raise Py4JJavaError(
py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.
: java.net.BindException: Can't assign requested address: Service 'sparkDriver' failed after 16 retries (on a random free port)! Consider explicitly setting the appropriate binding address for the service 'sparkDriver' (for example spark.driver.bindAddress for SparkDriver) to the correct binding address.
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:438)
	at sun.nio.ch.Net.bind(Net.java:430)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:225)
	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:562)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:260)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)

[2022-04-20 09:43:14 -0700] [1361] [INFO] Worker exiting (pid: 1361)
[2022-04-20 09:43:14 -0700] [1359] [INFO] Shutting down: Master
[2022-04-20 09:43:14 -0700] [1359] [INFO] Reason: Worker failed to boot.
Traceback (most recent call last):
  File "/Users/i29140/opt/anaconda3/envs/spark/bin/mlflow", line 8, in <module>
    sys.exit(cli())
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/mlflow/models/cli.py", line 65, in serve
    return _get_flavor_backend(
  File "/Users/i29140/opt/anaconda3/envs/spark/lib/python3.8/site-packages/mlflow/pyfunc/backend.py", line 197, in serve
    raise Exception(
Exception: Command '['bash', '-c', 'exec gunicorn --timeout=60 -b 0.0.0.0:15 -w 1 ${GUNICORN_CMD_ARGS} -- mlflow.pyfunc.scoring_server.wsgi:app']' returned non zero return code. Return code = 3
(spark) i29140@LVVCHD7QMD6R spark % mlflow models serve -m spark_model --no-conda --host 0.0.0.0 --port 15

--- a/mlflow/spark.py
+++ b/mlflow/spark.py
@@ -735,6 +735,9 @@ def _load_pyfunc(path):
             # In Spark 3.1 and above, we need to set this conf explicitly to enable creating
             # a SparkSession on the workers
             .config("spark.executor.allowSparkContext", "true")
+            # Bind "spark.driver.bindAddress" to 127.0.0.1 helps avoiding some local hostname
+            # related issues.
+            .config("spark.driver.bindAddress", "127.0.0.1")
             .master("local[1]")
             .getOrCreate()
         )

--- a/mlflow/spark.py
+++ b/mlflow/spark.py
@@ -735,8 +735,8 @@ def _load_pyfunc(path):
             # In Spark 3.1 and above, we need to set this conf explicitly to enable creating
             # a SparkSession on the workers
             .config("spark.executor.allowSparkContext", "true")
-            # Bind "spark.driver.bindAddress" to 127.0.0.1 helps avoiding some local hostname
-            # related issues.
+            # Binding "spark.driver.bindAddress" to 127.0.0.1 helps avoiding some local hostname
+            # related issues (e.g. https://github.com/mlflow/mlflow/issues/5733).
             .config("spark.driver.bindAddress", "127.0.0.1")
             .master("local[1]")
             .getOrCreate()

